Add SpringMVC Application files

Nikitosique opened the pull request:
Create application structure, add config files, add models files.
Add DAO and Mapper files for BodyType model.

Nikitosique committed:
Add SpringMVC Application files

Create application structure, add config files, add models files.
Add DAO and Mapper files for BodyType model.

Nikitosique commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,40 @@
+
+    public BodyType showById(int id) {
+        return jdbcTemplate.query("SELECT * FROM body_types WHERE id=?", new BodyTypeMapper(), id)
+                .stream().findAny().orElse(null);
```
Если запись не найдена, можно ли действовать следующим образом: 
1) возвращать нулл 
2) в контроллере просматривать возвращенный объект и если это нулл - показывать вью с ошибкой, если не нулл -  показывать вью с информацией об объекте

Или лучше выбрасывать исключение в таком случае?

AndryLat replied:
Spring JDBC templates throws back an EmptyResultDataAccessException exception if it doesn't find the data in the database.

AndryLat replied:
Also you should avoid to return null.

Nikitosique commented on src/main/java/dev/andrylat/carsharing/models/User.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public User(int discountCardId, String email, String password, String type) {
+        this.discountCardId = discountCardId;
```
Как реализовать композицию между юзером и скидочной картой?
Что-то вроде 
        this.discountCardId = new DiscountCard(discountValue, cardNumber).getId();

AndryLat replied:
it is not clear for me.Who is owner of relationship? I think it is "discount card". so, that objects should contain userId, it will be enough.

AndryLat commented on src/main/java/dev/andrylat/carsharing/config/MvcDispatcherSerlvetIntitializer.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Override
+    public void onStartup(ServletContext aServletContext) throws ServletException {
```
Is it really needed for you? (and next two methods)

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,40 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public List<BodyType> showAll() {
```
Show? where? on display?

"getAll"

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,40 @@
+                .stream().findAny().orElse(null);
+    }
+
+    public void add(BodyType bodyType) {
```
You will add new body type. How will you receive inserted data? for example, you should send added object to the front to show user this data and give him possibility to remove or update data.

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,40 @@
+        jdbcTemplate.update("INSERT INTO body_types VALUES(?)", bodyType.getName());
+    }
+
+    public void updateById(int id, BodyType updatedBodyType) {
```
Why do you need to set id here? May I take one from updatedBodyType?

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public void deleteById(int id) {
+        jdbcTemplate.update("DELETE FROM body_types WHERE id=?", id);
```
How will service know record was deleted or not?

AndryLat commented on src/main/java/dev/andrylat/carsharing/models/Car.java:
```diff
@@ -0,0 +1,70 @@
+    private int id;
+    private int modelId;
+    private int rentCostPerMin;
+    private String regPlate;
```
what is regPlate? what is "reg"?

AndryLat commented on src/main/java/dev/andrylat/carsharing/models/CarModel.java:
```diff
@@ -0,0 +1,89 @@
+    public CarModel() {
+    }
+
+    public CarModel(int bodyId, int brandId, int fuelId, String name, double engineDisplacement, String gearboxType, int productionYear) {
```
Is really needed? So many parameters there...

AndryLat commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+# Data Source Configuration
+spring.datasource.driver-class-name=org.postgresql.Driver
```
Never put credentials on Github!

Nikitosique committed:
Improve model files, add test files

Add small improvements to model files, add test files for DAO class,
add initial scripts for tests, add dependencies to pom.xml, delete application.properties

Nikitosique committed:
Delete application.properties

Nikitosique committed:
Delete view file

Nikitosique committed:
Rename local variables

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,87 @@
+        return jdbcTemplate.query("SELECT * FROM body_types WHERE id=?", new BodyTypeMapper(), id)
+                .stream()
+                .findAny()
+                .orElseThrow(() -> new EmptyResultDataAccessException("Could not find record with id " + id, 1));
```
at will be thrown before, as I remember

Nikitosique replied:
When this exception is throwing, my custom message appears (Could not find record with id). 
Does it mean, that orElseThrow method handles that exception?
Or I should use try-catch block?

AndryLat replied:
It is redundant actions. 
We can leave only jdbcTemplate.query("SELECT * FROM body_types WHERE id=?", new BodyTypeMapper(), id), set "queryObject", because we know that only one object must be returned.

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,87 @@
+        jdbcTemplate.update(
+                connection -> {
+                    PreparedStatement statement =
+                            connection.prepareStatement("INSERT INTO body_types(name) VALUES(?)", new String[]{"id"});
```
queries can be moved to constants.

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,87 @@
+
+        Optional<Number> insertedRecordIdOptional = Optional.ofNullable(keyHolder.getKey());
+        int insertedRecordId = (int) insertedRecordIdOptional
+                .orElseThrow(() -> new NullPointerException("Data insertion has failed! Couldn't get inserted record!"));
```
never throw NPE. Also NPE here is no correct at all.

Nikitosique replied:
What exception should I use instead?
RecordNotFoundException?

AndryLat replied:
yes

AndryLat commented on src/main/java/dev/andrylat/carsharing/dao/BodyTypeDAO.java:
```diff
@@ -0,0 +1,87 @@
+        return deletedRowsNumber > 0;
+    }
+
+    public boolean deleteAll() {
```
bad method, better to remove. In real life it is not a business case.

Nikitosique committed:
Add small improvements to application

Add RecordNotFoundException class, remove deleteAll method,
add queries to class constants.

Nikitosique committed:
Rename constants, add .properties file for tests

Nikitosique committed:
Add CarModelDAO

Add CarModelDAO files, add tests for CarModelDAO

Nikitosique committed:
Add small improvements

AndryLat commented on pom.xml:
```diff
@@ -0,0 +1,73 @@
+    <artifactId>CarSharing</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <name>CarSharing</name>
+    <description>Demo project for Spring Boot</description>
```
Or remove or change.

AndryLat commented on pom.xml:
```diff
@@ -0,0 +1,73 @@
+
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
```
why is it not in test scope?

AndryLat commented on src/main/java/dev/andrylat/carsharing/config/SpringConfig.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Bean
+    public DataSource dataSource() {
```
https://www.baeldung.com/properties-with-spring

Nikitosique committed:
Add DAO-classes and mappers for application models

Add DAO-classes, add mappers, add embedded PostgreSQL data source for
tests.

Nikitosique committed:
Add pagination for BodyTypeDAO

Nikitosique committed:
Add small improvements to BodyTypeDAO

Nikitosique committed:
Fix insertedRecordId type mismatch
